### PR TITLE
Update master

### PR DESCRIPTION
**Pull request by johancx**
# Changes :gem:
* DB redesign fro player and game
* Score supplied by front end
* Labels supplied by database
  * Labels are now in Norwegian
* replaced deprecated flask jsonify with standard python json

### CODE REVIEW

File: src/test/test_api.py [hunk 1]
```diff
@@ -223,18 +224,31 @@ def construct_path(dir_list):
...
     assert(isinstance(response["total"], list))
     assert(isinstance(response["daily"][0], dict))
     assert(isinstance(response["total"][0], dict))
+def write_to_db():
+    with api.app.app_context():
+        for i in range(5):
+            result = models.insert_into_scores(
+                "Test User", 10 + i, datetime.date.today() - datetime.timedelta(days=i))
+            assert result
+'''
```
Comment (andreacx) [inline src/test/test_api.py:254]: This test has to be fixed

File: src/test/test_db.py [hunk 2]
```diff
@@ -12,9 +12,11 @@
 from werkzeug import exceptions as excp
 
 token = uuid.uuid4().hex
+game_id = uuid.uuid4().hex
 labels = "label1, label2, label3"
-play_time = 21.0
+play_time = 11.0
 start_time = time.time()
```
Comment (andreacx) [inline src/test/test_db.py:18]: remove start_time. Old variable

File: src/utilities/dict_eng_to_nor.csv [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+pizza,pizza
+police car,politibil
+screwdriver,skrutrekker
+sheep,sau
+snail,snegle
+suitcase,koffert
+tractor,traktor
+watermelon,vannmelon
+wheel,hjul
+wristwatch,klokke
```
Comment (andreacx) [inline src/utilities/dict_eng_to_nor.csv:48]: no new line at end of file?

File: src/webapp/api.py [hunk 4]
```diff
@@ -101,33 +102,35 @@ def classify():
...
-
     best_guess, certainty = classifier.predict_image(image)
     # use token submitted by player to find game
     token = request.values["token"]
     # Get time from POST request
-    time_used = float(request.values["time"])
+    time_left = float(request.values["time"])
     # Get label for game session
-    game = models.get_record_from_game(token)
+    player_in_game = models.get_record_from_player_in_game(token)
```
Comment (andreacx) [inline src/webapp/api.py:111]: player_in_game or player?

File: src/webapp/models.py [hunk 5]
```diff
@@ -4,30 +4,25 @@
...
-    """
-    pass
-
-
 class Games(db.Model):
     """
        This is the Games model in the database. It is important that the
        inserted values match the column values. Token column value cannot
        be String when a long hex is given.
     """
```
Comment (andreacx) [inline src/webapp/models.py:22]: remove new line

File: src/webapp/models.py [hunk 6]
```diff
@@ -37,12 +32,33 @@ class Scores(db.Model):
         inserted values match the column values.
     """
```
Comment (andreacx) [inline src/webapp/models.py:34]: remove new line

File: src/webapp/models.py [hunk 7]
```diff
@@ -128,53 +183,57 @@ def update_game(token, session_num, play_time):
         raise Exception("Couldn't update game.")
```
Comment (andreacx) [inline src/webapp/models.py:183]: Time for deleting this function? Its been replaced by the one belove

File: src/webapp/models.py [hunk 8]
```diff
@@ -228,13 +287,67 @@ def drop_table(table):
...
+        if os.path.exists(filepath):
+            # clear table
+            Labels.query.delete()
+            db.session.commit()
+            with open(filepath) as csvfile:
+                try:
+                    readCSV = csv.reader(csvfile, delimiter=',')
+
+                    for row in readCSV:
+                        insert_into_labels(row[0], row[1])
```
Comment (andreacx) [inline src/webapp/models.py:303]: new line after for loop

File: src/webapp/models.py [hunk 9]
```diff
@@ -228,13 +287,67 @@ def drop_table(table):
...
+    """
+    try:
+        # read all english labels in database
+        labels = Labels.query.all()
+        english_labels = [str(label.english) for label in labels]
+        random_list = random.sample(english_labels, n)
+        return random_list
+
+    except Exception as e:
+        return Exception("Could not read " + str(e) + " random rows from Labels table")
```
Comment (andreacx) [inline src/webapp/models.py:340]: RETURN EXCEPTION?

File: src/webapp/models.py [hunk 10]
```diff
@@ -228,13 +287,67 @@ def drop_table(table):
...
-        return rows
-    elif table == "Scores":
-        rows = db.session.query(Scores).count()
-        return rows
+    try:
+        norwegian_word = Labels.query.get(english_label)
+        return norwegian_word
+
+    except AttributeError as e:
+        return AttributeError("Could not find translation in Labels table: " + str(e))
```
Comment (andreacx) [inline src/webapp/models.py:352]: RETURN EXCEPTION

File: src/webapp/models.py [hunk 11]
```diff
@@ -128,53 +182,57 @@ def update_game(token, session_num, play_time):
         raise Exception("Couldn't update game.")
 
 
-def delete_session_from_game(token):
+# ALTERNATIVE FUNC FOR UPDATE GAME TO ALSO WORK FOR MULTI
+def update_game_for_player(game_id, token, session_num, play_time):
+    """
+        Docstring.
```
Comment (andreacx) [inline src/webapp/models.py:188]: Change docstring to something more explainable

File: src/test/test_db.py [hunk 12]
```diff
@@ -140,17 +185,61 @@ def test_get_top_n_high_score_list_structure():
         assert "name" in player
 
 
-'''
-def test_clear_table():
+def test_get_iteration_name_is_string():
+    """
+        Tests if it's possible to get an iteration name from the database and the type is str
+    """
```
Comment (andreacx) [inline src/test/test_db.py:191]: too long :(

File: src/utilities/dict_eng_to_nor.csv [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+hammer,hammer
+hand,hånd
+helicopter,helikopter
+horse,hest
+hospital,sykehus
+key,nøkkel
+lightning,lyn
+mermaid,havfrue
+mountain,fjell
+ocean,sjø
```
Comment (joaaax) [inline src/utilities/dict_eng_to_nor.csv:35]: hav?

File: src/test/test_db.py [hunk 14]
```diff
@@ -140,17 +185,61 @@ def test_get_top_n_high_score_list_structure():
...
+    with api.app.app_context():
+        for i in range(5):
+            result = models.get_n_labels(i)
+            assert len(result) == i
+
+
+def test_get_n_labels_bad_reqeust():
+    """
+        Test that get_n_labels raises exeption if n is larger than number of labels
+    """
```
Comment (andreacx) [inline src/test/test_db.py:211]: too long agian :((

File: src/utilities/setup.py [hunk 15]
```diff
@@ -2,12 +2,17 @@
...
 
-# USED BY CUSTOM VISION
+
+# number of players in overall high score top list
+TOP_N = 10
+# Total number of games
+NUM_GAMES = 3
+# certainties from costum vision lower than this -> haswon=False
+CERTAINTY_THRESHOLD = 0.5
 # custom vision can't have more than 10 iterations at a time, if more classifier.py will delete the oldest iteration
```
Comment (andreacx) [inline src/utilities/setup.py:12]: too long...

File: src/utilities/dict_eng_to_nor.csv (hunk 3 above)
Comment (joaaax) [inline src/utilities/dict_eng_to_nor.csv:48]: Armbåndsur?

File: src/webapp/api.py [hunk 16]
```diff
@@ -101,33 +100,35 @@ def classify():
...
+        and best_certainty >= CERTAINTY_THRESHOLD
     )
     # End game if player win or loose
-    if has_won or time_used >= TIME_LIMIT:
+    if has_won or time_left <= 0:
         # save image in blob storage
         storage.save_image(image, label)
         # Get cumulative time
-        cum_time = game.play_time + time_used
+        cum_time = player.play_time + time_left
```
Comment (andreacx) [inline src/webapp/api.py:125]: Play_time should be changed to score or something

File: src/webapp/models.py [hunk 17]
```diff
@@ -54,29 +75,33 @@ def create_tables(app):
...
-            game = Games(token=token, labels=labels,
-                         play_time=play_time, date=date)
+            game = Games(game_id=game_id, labels=labels, date=date)
             db.session.add(game)
             db.session.commit()
             return True
-        except DataBaseException:
-            raise DataBaseException("Could not insert into games")
+        except Exception as e:
+            raise Exception("Could not insert into games :" + str(e))
```
Comment (andreacx) [inline src/webapp/models.py:99]: sry my fault... Should be: ("...games: ") not ("...games :")

File: src/webapp/models.py [hunk 18]
```diff
@@ -128,53 +226,66 @@ def update_game(token, session_num, play_time):
...
-        game = Games.query.filter_by(token=token).first()
+        game = Games.query.get(game_id)
+        db.session.query(PlayerInGame).filter(
+            PlayerInGame.game_id == game_id
+        ).delete()
         db.session.delete(game)
         db.session.commit()
-        return "Record connected to " + token + " deleted."
-    except AttributeError:
+        return "Record deleted."
```
Comment (andreacx) [inline src/webapp/models.py:259]: unnecessary return?

File: src/utilities/setup.py [hunk 19]
```diff
@@ -2,12 +2,17 @@
...
 CV_MAX_ITERATIONS = 10
 # can't upload more than 64 images at a time, if more
 CV_MAX_IMAGES = 64
-
 # USED BY API
 LABELS = [
```
Comment (jennybhel) [inline src/utilities/setup.py:17]: fjerne dette?

File: src/webapp/models.py [hunk 20]
```diff
@@ -228,13 +349,69 @@ def drop_table(table):
...
+
+                    for row in readCSV:
+                        insert_into_labels(row[0], row[1])
+                except AttributeError as e:
+                    raise AttributeError(
+                        "Could not insert into games: " + str(e)
+                    )
+
+        else:
+            raise AttributeError("File path not found")
```
Comment (andreacx) [inline src/webapp/models.py:373]: include builtin exception?
+ str(e)

**Pull request merged**